src/memtest.rs

```diff
@@ -0,0 +1,486 @@
+/// pair, write a random value. After all locations are written, read and compare the two halves.
+#[tracing::instrument(skip_all)]
+pub fn test_random_val(
+    memory: &mut [usize],
+    mut timeout_checker: TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    let (first_half, second_half) = split_slice_in_half(memory)?;
+    let expected_iter =
+        u64::try_from(first_half.len() * 2).context("Total number of iterations overflowed")?;
+    timeout_checker.init(expected_iter);
+
+    for (first_ref, second_ref) in first_half.iter_mut().zip(second_half.iter_mut()) {
+        timeout_checker.check()?;
+        let val = random();
+        write_volatile_safe(first_ref, val);
+        write_volatile_safe(second_ref, val);
+    }
+
+    compare_regions(first_half, second_half, &mut timeout_checker)
+}
+
+/// Reset all bits in given memory to 1s. Split given memory into two halves and iterate through
+/// memory locations in pairs. For each pair, write the XOR result of a random value and the value
+/// read from the location. After all locations are written, read and compare the two halves.
+#[tracing::instrument(skip_all)]
+pub fn test_xor(
+    memory: &mut [usize],
+    timeout_checker: TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    test_two_regions(memory, timeout_checker, std::ops::BitXor::bitxor)
+}
+
+/// Reset all bits in given memory to 1s. Split given memory into two halves and iterate through
+/// memory locations in pairs. For each pair, write the result of subtracting a random value from
+/// the value read from the location. After all locations are written, read and compare the two
+/// halves.
+#[tracing::instrument(skip_all)]
+pub fn test_sub(
+    memory: &mut [usize],
+    timeout_checker: TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    test_two_regions(memory, timeout_checker, usize::wrapping_sub)
+}
+
+/// Reset all bits in given memory to 1s. Split given memory into two halves and iterate through
+/// memory locations in pairs. For each pair, write the result of multiplying a random value with
+/// the value read from the location. After all locations are written, read and compare the two
+/// halves.
+#[tracing::instrument(skip_all)]
+pub fn test_mul(
+    memory: &mut [usize],
+    timeout_checker: TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    test_two_regions(memory, timeout_checker, usize::wrapping_mul)
+}
+
+/// Reset all bits in given memory to 1s. Split given memory into two halves and iterate through
+/// memory locations in pairs. For each pair, write the result of dividing the value read from the
+/// location with a random value. After all locations are written, read and compare the two halves.
+#[tracing::instrument(skip_all)]
+pub fn test_div(
+    memory: &mut [usize],
+    timeout_checker: TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    test_two_regions(memory, timeout_checker, usize::wrapping_div)
```

Comment:
Unfortunately, here you need an ever-so-slightly more complicated mixing function.

```Rust
test_two_regions(memory, timeout_checker, |n, d| n.wrapping_div(usize::max(d, 1)))
```